Issue #16968 : Convert ALTS handshaker client code from C to C++

RedMarcher opened the pull request:
I will say right away that I'm new to contributing to gRPC and as such choose this issue because it seemed like a good introduction. 

I removed typedefs from the code, since they seem to be a typical holdover from C code, and not really necessary for the updated code. I also modified one of the const values to be a constexpr. Lastly, I added in a missing comment.

Please let me know if this is properly addressing the issue, I look forward to receiving feedback and learning lots of new things!

RedMarcher commented:
And BTW, if someone could explain to me how to add a label for Mergeable to be appeased, I'd appreciate it.

RedMarcher commented:
Alright, I started the process of putting the old struct based systems into classes. I would appreciate some feedback as to whether or not I did that correctly. Don't worry, I know I have to change the other parts as well, I just want to make sure my format and code style is all correct.

RedMarcher commented:
Well, since it took a while for anyone to respond on this particular issue, I asked a Googler from another team to take a quick look at what I had written so far, and he said it was OK, so I continued work.

I realize now why this issue has been on the backburner for a while, so much code to rewrite. There's seemingly a lot of code in the handshaker client cc file that isn't actually a part of handshaker? It's wrapped up in separate namespaces. 

Also, I noticed the "virtual functions" were also defined in the cc file. I'm a little confused, I thought virtual functions only got defined once inheritance occurred?

RedMarcher commented:
As for the other functions(methods) which were not defined in the h file but are existant in the cc file, am I supposed to assume these are meant to be enwrapped within the class as well?

gtcooke94 commented:
Hey, sorry for the delay, and thanks for taking a go at this!

Sorry I didn't get to take a look at this earlier last week. You're right that it's a lot of code to re-write into idiomatic C++, and it looks like you're making good progress!

As a general tip, everything that is a method like `tsi_result alts_handshaker_client_next(alts_handshaker_client* client, grpc_slice* bytes_received)` that takes a pointer to (essentially) itself as the first arg are all functions that could be on a class of that type instead, then they don't need the first pointer arg because it'll be it's own instance of the class - does that make sense?

In terms of things defined in `.h` vs. `.cc` - `.h` files contain the definitions that can be `#included` in another file and used elsewhere, so that represents the API for that subset of code. Then internal helpers/abstractions that wouldn't need to be used outside of the given file can remain in the `.cc` file.

RedMarcher commented:
Looks like there's a lot of function changes removing vtable and self references as well. Seems like this is quite the non-trivial rewrite.

I assume alts_grpc_handshaker_client_create is meant to be a sort of constructor? Behavior wise, does it have to return a pointer, or can it just be implemented as a constructor?

RedMarcher commented:
> Hey, sorry for the delay, and thanks for taking a go at this!
> 
> Sorry I didn't get to take a look at this earlier last week. You're right that it's a lot of code to re-write into idiomatic C++, and it looks like you're making good progress!
> 
> As a general tip, everything that is a method like `tsi_result alts_handshaker_client_next(alts_handshaker_client* client, grpc_slice* bytes_received)` that takes a pointer to (essentially) itself as the first arg are all functions that could be on a class of that type instead, then they don't need the first pointer arg because it'll be it's own instance of the class - does that make sense?
> 
> In terms of things defined in `.h` vs. `.cc` - `.h` files contain the definitions that can be `#included` in another file and used elsewhere, so that represents the API for that subset of code. Then internal helpers/abstractions that wouldn't need to be used outside of the given file can remain in the `.cc` file.

You mean all the methods that take in a pointer of the handshaker_client can be replaced with "this"? You're certain that none of those methods are meant to take in a pointer to a potentially different handshaker_client? I haven't gone super in depth into the actual methods themselves yet. I noticed some seemingly global functions being used.

gtcooke94 commented:
> I assume alts_grpc_handshaker_client_create is meant to be a sort of constructor? Behavior wise, does it have to return a pointer, or can it just be implemented as a constructor?

My initial thought is to make it a static factory function. And these things can also be returning `std::shared_ptr` or `std::unique_ptr` depending on the use, rather than raw pointers - smart pointers are a wonderful thing that we gain moving to modern C++.

> You mean all the methods that take in a pointer of the handshaker_client can be replaced with "this"? You're certain that none of those methods are meant to take in a pointer to a potentially different handshaker_client? 

Correct, I'm pretty certain. In C89 there isn't a concept of a member function on a class (because there are no classes), so something like `<struct_name>_<function>(ptr_to_struct*, ...)` can be more cleanly expressed as `ClassName::Function` in C++.

e.g. `void alts_handshaker_client_handle_response(alts_handshaker_client* c,
                                            bool is_ok)` -> `AltsHandshakerClient::HandleResponse(bool is_ok)`

RedMarcher commented:
I'm noticing quite a lot of statics in this class. Is this class ever meant to have multiple objects at once, or is the entire thing just meant to exist once?

gtcooke94 commented:
The functions can be static in C89 because they aren't associated with an object/struct - just because they are static currently does not mean they should be static in idiomatic C++.

RedMarcher commented:
alts_grpc_handshaker_client* client =
      reinterpret_cast<alts_grpc_handshaker_client*>(c);

Why does this code show up so much? Isn't this literally casting a pointer as the same kind of pointer? Is this some more C-style magic?

RedMarcher commented:
alts_handshaker_client_destroy()
This function seems a lot like a destructor

Like you already mentioned, there's a factory method in this class.

Instead of the traditional constructor and destructor, do you instead want objects from this class to be created using a static factory method and released using a static destructor method?

gtcooke94 commented:
It is a destructor for the C89 struct - so, in conversion to C++, it should be a destructor

In terms of using static factories vs. constructors, it depends. Static factories let us do initialization and use private constructors that we may not want accessible otherwise.

RedMarcher commented:
> It is a destructor for the C89 struct - so, in conversion to C++, it should be a destructor
> 
> In terms of using static factories vs. constructors, it depends. Static factories let us do initialization and use private constructors that we may not want accessible otherwise.

If I have the factory method return a unique_ptr as you suggested, do we even need a destructor function anymore? Doesn't unique_ptr handle deletion for us?
Also, what's the difference between destroy and shutdown exactly?

RedMarcher commented:
Alright, I think I'm getting close to wrapping this up. I decided to keep the self-referential pointers for the public-facing methods for now just for the sake of maximizing compatibility with preexisting code. All I have to do now is to change the old name to the new name of the class, and I think it should be good for now.

How would I go about checking for compile time errors with this code? Is there a specific testing process that Google uses?

RedMarcher commented:
Alright, I renamed all the references and added scopes where necessary for the new class.

I have one final question. For the public methods, is there a way to know which function the old vtable was trying to call? Names like "shutdown" or "next" aren't very descriptive.

gtcooke94 commented on src/core/tsi/alts/handshaker/alts_handshaker_client.h:
```diff
@@ -34,131 +34,209 @@
+  ///
+  ///- client: ALTS handshaker client instance.
+  ///
+  static void alts_handshaker_client_shutdown(AltsHandshakerClient* client);
```
I think we should change these from being static and self referentials, the following should be functionally identical.

```
AltsHandshakerClient* client = <somehow created>
alts_handshaker_client_shutdown(client);
client->Shutdown(); // (a new instance method)
```

RedMarcher replied:
I was originally going to do that, but I realized that would involve changing a lot of code outside this class. Maybe I can open up another PR for this afterward. For this PR, I decided to keep those as sort of wrapper functions.

RedMarcher replied:
On second thought, I realize that won't work because the static function won't be able to call private methods or access private variables anyways. Alright, I will just bite the bullet and reformat the other code as well.

gtcooke94 commented on src/core/tsi/alts/handshaker/alts_handshaker_client.h:
```diff
@@ -34,131 +34,209 @@
+  ///
+  /// It returns TSI_OK on success and an error status code on failure.
+  ///
+  static tsi_result alts_handshaker_client_next(AltsHandshakerClient* client, grpc_slice* bytes_received);
```
Throughout the PR, can we use modern C++ casing? i.e. `ClientNext`

gtcooke94 commented on test/core/tsi/alts/handshaker/alts_handshaker_client_test.cc:
```diff
@@ -228,8 +229,9 @@ static grpc_call_error check_server_start_success(grpc_call* /*call*/,
   upb::Arena arena;
-  alts_handshaker_client* client =
-      static_cast<alts_handshaker_client*>(closure->cb_arg);
+  AltsHandshakerClient* client =
```
Throughout, can we refactor to using smart pointers (`std::unique_ptr<AltsHandshakerClient>`) instead of raw pointers?

gtcooke94 reviewed (commented):
This is looking really good, thanks for the PR! I left a couple of comments that apply broadly to the style of the new code.

As for testing, when it's looking good, we can move it out of `draft` state and then start running a large test suite against it via github CI

In the meanwhile, locally testing with `bazel` should be pretty straight-forward if you haven't been doing it yet

RedMarcher commented:
I made some more progress. I guess there's no avoiding it, I think in addition to editing the file I originally set out to change, I unfortunately will have to also change a lot of the logic for the code that calls AltsHandshakerClient as well.

The large amount of casting definitely doesn't help my efforts, but I'll gradually fix everything.

RedMarcher commented:
Ok, so I'm a little confused how to build and run tests. I'm using a codespace, and I followed the steps to set up the environment, including importing the submodules.

Now do I just type in: python tools/run_tests/run_tests.py -l c++ --build_only
to the console in order to build? Then to test I just type that in without the --build_only?

gtcooke94 commented:
This is also considered C-Core, so we need to run the `c` and `c++` tests - that should work

You should also be able to run `bazel test //path/to:target` to test as well

RedMarcher commented:
I just wanted to let you know I'm still working on this. Sorry this is taking longer than expected, it's taking a while to change everything that needs to be changed after the rewrite.

RedMarcher commented:
Oh man, that took quite a while, but I think it's ready for official code review now.

I was, unfortunately, unable to run any of the tests locally due to compiler issues that I'm pretty certain are due to using cmake rather than bazel. I did make sure to fix all the obvious errors that popped up though.

Thanks for letting me work on this.

gtcooke94 commented:
Thanks for all the hard work! Sorry I've been busy since the new year - I'll get to this soon

RedMarcher commented:
> Thanks for all the hard work! Sorry I've been busy since the new year - I'll get to this soon

You still with me? I asked a Googler from another team and he said your team was really overworked recently. I'm a bit worried about their being a conflict because I changed quite a few files.
